src/Utilities/PointerVector.hpp

```diff
@@ -85,6 +85,191 @@ template <typename T>
+  };
+};
+
+template <typename VT, bool TF, typename Abs, typename Power, typename Root>
+inline decltype(auto) norm_backend(const DenseVector<VT, TF>& dv, Abs abs,
+                                   Power power, Root root, FalseType) noexcept {
+  using CT = CompositeType_<VT>;
+  using ET = ElementType_<VT>;
+  using RT = decltype(evaluate(root(std::declval<ET>())));
+
+  if ((~dv).size() == 0UL)
+    return RT();
+
+  CT tmp(~dv);
+
+  const size_t N(tmp.size());
+
+  ET norm(power(abs(tmp[0UL])));
+  size_t i(1UL);
+
+  for (; (i + 4UL) <= N; i += 4UL) {
+    norm += power(abs(tmp[i])) + power(abs(tmp[i + 1UL])) +
+            power(abs(tmp[i + 2UL])) + power(abs(tmp[i + 3UL]));
+  }
+  for (; (i + 2UL) <= N; i += 2UL) {
+    norm += power(abs(tmp[i])) + power(abs(tmp[i + 1UL]));
+  }
+  for (; i < N; ++i) {
+    norm += power(abs(tmp[i]));
+  }
+
+  return evaluate(root(norm));
+}
+
+template <typename VT, bool TF, typename Abs, typename Power, typename Root>
+inline decltype(auto) norm_backend(const DenseVector<VT, TF>& dv, Abs abs,
+                                   Power power, Root root, TrueType) noexcept {
+  using CT = CompositeType_<VT>;
+  using ET = ElementType_<VT>;
+  using RT = decltype(evaluate(root(std::declval<ET>())));
+
+  enum : size_t { SIMDSIZE = SIMDTrait<ET>::size };
+
+  if ((~dv).size() == 0UL)
+    return RT();
+
+  CT tmp(~dv);
+
+  const size_t N(tmp.size());
+
+  constexpr bool remainder(!usePadding || !IsPadded<VT>::value);
+
+  const size_t ipos((remainder) ? (N & size_t(-SIMDSIZE)) : (N));
+  BLAZE_INTERNAL_ASSERT(!remainder || (N - (N % SIMDSIZE)) == ipos,
+                        "Invalid end calculation");
+
+  SIMDTrait_<ET> xmm1, xmm2, xmm3, xmm4;
+  size_t i(0UL);
+
+  for (; (i + SIMDSIZE * 3UL) < ipos; i += SIMDSIZE * 4UL) {
+    xmm1 += power(abs(tmp.load(i)));
+    xmm2 += power(abs(tmp.load(i + SIMDSIZE)));
+    xmm3 += power(abs(tmp.load(i + SIMDSIZE * 2UL)));
+    xmm4 += power(abs(tmp.load(i + SIMDSIZE * 3UL)));
+  }
+  for (; (i + SIMDSIZE) < ipos; i += SIMDSIZE * 2UL) {
+    xmm1 += power(abs(tmp.load(i)));
+    xmm2 += power(abs(tmp.load(i + SIMDSIZE)));
+  }
+  for (; i < ipos; i += SIMDSIZE) {
+    xmm1 += power(abs(tmp.load(i)));
+  }
+
+  ET norm(sum(xmm1 + xmm2 + xmm3 + xmm4));
+
+  for (; remainder && i < N; ++i) {
+    norm += power(abs(tmp[i]));
+  }
+
+  return evaluate(root(norm));
+}
+
+template <typename VT, bool TF, typename Abs, typename Power, typename Root>
+decltype(auto) norm_backend(const DenseVector<VT, TF>& dv, Abs abs, Power power,
+                            Root root) {
+  return norm_backend(~dv, abs, power, root,
+                      Bool<DVecNormHelper<VT, Abs, Power>::value>());
+}
+
+template <typename VT, bool TF>
+decltype(auto) l1Norm(const DenseVector<VT, TF>& dv) noexcept {
+  return norm_backend(~dv, Abs(), Noop(), Noop());
+}
+
+template <typename VT, bool TF>
+inline decltype(auto) l2Norm(const DenseVector<VT, TF>& dv) noexcept {
```

Comment:
I'd prefer this to not be operable on complex types, because the complex version will need an abs in the first slot or a different power operation in order to make sense. I'd be happy with it either being SFINAE here or disallowed via the blaze map traits. The type trait route might be preferable if you're trying to minimally change the code imported from Blaze.

Reply:
Seems reasonable. I'd prefer to have that be done as a trait if possible since I'd like to have Blaze behavior be version independent. Do you have a suggestion on how to do this with the traits?

Reply:
well, shoot. I thought I did, but looking at this more carefully, there isn't actually a blaze struct for the norm function -- usually the dispatch here ends up going through a map or some such internal blaze thing, which gets the 'result type' from the type trait struct.
Alright, my suggestion is to try the following, and if it doesn't work, just leave it as is. Feel free to just make this an issue and assign it to me if you want to get this merged quickly or are unsatisfied with the below solution.
Put in `ComplexDataVector.hpp` a partial template specialization to the helper struct:
```
namespace blaze {
template <typename Abs, typename Power>
struct DVecNormHelper<ComplexDataVector, Abs, Power> {
};
}
```
I think this should make the bool lookup fail on line 262 and therefore not compile l2norm of a ComplexDataVector. Not as nice as I had hoped...

Reply:
I had to do something slightly different. This works for `l2Norm(vector)` but fails for `l2Norm(vector * vector)` (and probably other expressions). If desired we could add other specializations that can maybe deal with this more generally.

Reply:
I'm happy with this for now, there might be some template tricks we can pull to select anything for which the `ResultType` is `ComplexDataVector` in that specialization. Thanks for putting this in.